Strip less aggressively to make the embedded hash survive

rschu1ze opened the pull request:
- It was noticed that in (*), the crashstack says "There is no information about the reference checksum."

- The binaries are pulled via docker hub and upon inspection they indeed lack the hash embedded as ELF section ".note.ClickHouse.hash" in the clickhouse binary. This is weird because docker hub images are "official" builds which should trigger the hash embedding.

- Turns out that the docker hub binaries are also stripped which removed the hash section. This PR moves the hash into ELF section '.clickhouse.hash' (so we can continue to strip '.comment' and '.note') and instructs stripping to spare '.clickhouse.hash'.

(*) https://playground.lodthe.me/ba75d494-95d1-4ff6-a0ad-60c138636c9b

### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix overly aggressive stripping which removed the embedded hash required for checking the consistency of the executable

rschu1ze committed:
Strip less aggressively to make the embedded hash survive

- It was noticed that in (*), the crashstack says "There is no
  information about the reference checksum."

- The binaries are pulled via docker hub and upon inspection they indeed
  lack the hash embedded as ELF section ".note.ClickHouse.hash" in the
  clickhouse binary. This is weird because docker hub images are
  "official" builds which should trigger the hash embedding.

- Turns out that the docker hub binaries are also stripped which was too
  aggressive. We now no longer remove sections ".comment" and ".note"
  which are anyways only 140 bytes in size, i.e. binary size still goes
  down (on my stystem) from 2.1 GB to 0.47 + 0.40 GB binary + dbg info.

(*) https://playground.lodthe.me/ba75d494-95d1-4ff6-a0ad-60c138636c9b

Felixoid commented on cmake/strip_binary.cmake:
```diff
@@ -21,7 +21,7 @@ macro(clickhouse_strip_binary)
        COMMAND "${OBJCOPY_PATH}" --only-keep-debug --compress-debug-sections "${STRIP_DESTINATION_DIR}/bin/${STRIP_TARGET}" "${STRIP_DESTINATION_DIR}/lib/debug/bin/${STRIP_TARGET}.debug"
        COMMAND chmod 0644 "${STRIP_DESTINATION_DIR}/lib/debug/bin/${STRIP_TARGET}.debug"
-       COMMAND "${STRIP_PATH}" --remove-section=.comment --remove-section=.note "${STRIP_DESTINATION_DIR}/bin/${STRIP_TARGET}"
+       COMMAND "${STRIP_PATH}" "${STRIP_DESTINATION_DIR}/bin/${STRIP_TARGET}"
```
I'd leave `--remove-section=.comment` to have it close to a former debian stripper

rschu1ze replied:
What do you mean by "former debian stripper"?

(Another approach would be to embed the hash as ".ClickHouse" (instead of ".note.ClickHouse.hash") and to continue remove ".note" and ".comment". But I did not really understand what ".note" and ".comment" were used for except the hash.)

Felixoid replied:
We more or less reimplemented the debian debug symbols stripper in cmake, see https://github.com/ClickHouse/ClickHouse/pull/35196

Maybe, using `.ClickHouse.hash` or `.diagnostics.ClickHouse.hash` is a good idea, but it's worth reading if it's fine.

rschu1ze replied:
Stripping in Debian: www.debian.org/doc/debian-policy/ch-files.html

So indeed, they explicitly remove `.note` and `.comment`. The real question is what `.note` and `.comment` actually store. These sections are non-standard and I found only vague information, e.g.
- https://lists.debian.org/debian-policy/1998/03/msg00256.html - a 1998 (!) mailthread
- https://jerry153fish.github.io/2017/10/11/ELF-format.html
- https://stackoverflow.com/questions/9401217/is-the-elf-notes-section-really-needed

In sum, they store the compiler version information or placeholders which seem to be non-mandatory for debugging. Also in some cases, both sections can be quite large, however, in ClickHouse, I found them actually very small. But maybe this also depends on the compiler.

Anyways, perhaps it's a good thing to stay compatible with Debian as long as we document it. Will push a follow-up in this PR.

rschu1ze replied:
And another LOL:
```
./clickhouse hash-binary  # returns C66066CFE511A7C34A9A0C6281F44C21

# add the hash to the binary (happens for official builds)
objcopy --add-section .ClickHouse.hash=hash clickhouse

./clickhouse hash-binary  # returns 37337872D77DD22744223DDB5E464913

./clickhouse server
[...]
<Error> Application: DB::Exception: Calculated checksum of the ClickHouse binary (37337872D77DD22744223DDB5E464913) does not correspond to the reference checksum stored in the binary (C66066CFE511A7C34A9A0C6281F44C21). It may indicate one of the following: - the file /home/ubuntu/repo/ClickHouse1/build/programs/clickhouse-bak was changed just after startup; - the file /home/ubuntu/repo/ClickHouse1/build/programs/clickhouse-bak is damaged on disk due to faulty hardware; - the loaded executable is damaged in memory due to faulty hardware; - the file /home/ubuntu/repo/ClickHouse1/build/programs/clickhouse-bak was intentionally modified; - logical error in code.
[...]
```
I really wonder if that hashsum feature was ever verified to work ;-)

alexey-milovidov replied:
Yes, it is verified to work.

rschu1ze committed:
Move binary hash to ELF section ".ClickHouse.hash"

rschu1ze committed:
Log official/non-official build and integrated/separate debug symbols

The goal is to find out why some of the binaries with official name
(BuilderDebRelease, BuilderBinRelease) produced by CI still contain no
hash section.

(also, strip and objcopy are mandatory tools and we don't need to check
for their existence at this point)

Felixoid commented on CMakeLists.txt:
```diff
@@ -244,16 +244,18 @@ endif ()
 # Only for official builds, because adding a section can be time consuming (rewrite of several GB).
 # And cross compiled binaries are not supported (since you cannot execute clickhouse hash-binary)
-if (OBJCOPY_PATH AND CLICKHOUSE_OFFICIAL_BUILD AND (NOT CMAKE_TOOLCHAIN_FILE OR CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-x86_64.cmake$"))
+if (CLICKHOUSE_OFFICIAL_BUILD AND (NOT CMAKE_TOOLCHAIN_FILE OR CMAKE_TOOLCHAIN_FILE MATCHES "linux/toolchain-x86_64.cmake$"))
```
we support aarch64 official builds too

Felixoid commented on CMakeLists.txt:
```diff
@@ -244,16 +244,18 @@ endif ()
-        set(STRIPPED_BINARIES_OUTPUT "stripped" CACHE STRING "A separate directory for stripped information")
-    endif()
+# Optionally split binaries and debug symbols.
+option(INSTALL_STRIPPED_BINARIES "Split binaries and debug symbols" OFF)
```
I'm pretty sure it will break horribly =)

Felixoid commented on CMakeLists.txt:
```diff
@@ -244,16 +244,18 @@ endif ()
 endif ()
 
-# Allows to build stripped binary in a separate directory
-if (OBJCOPY_PATH AND STRIP_PATH)
```
Removing it may break in funny places. This part has been adjusted very accurately for long time.

rschu1ze replied:
You found the right reference already 😄

About: objcopy/strip: In general, I think it is worse to silently skip the initialization of split build (because we are on a hypothetical platform without objcopy) than it is to do the initialization and fail with a weird message later on. But that is all theory.

The main point of the commit is to prove what really happens in the CI builds. I looked at some generated binaries (stored at Amazon) and they had unexpected ELF sections.

Felixoid commented:
Let's take a look on https://github.com/ClickHouse/ClickHouse/commit/88dfcaa8923b1b7fadef07127904366a9e74d0f8/checks/6869377705/logs from https://github.com/ClickHouse/ClickHouse/runs/6869377705?check_suite_focus=true

```
2022-06-13T14:34:47.2136082Z INFO:root:Going to run packager with cd /home/ubuntu/actions-runner/_work/_temp/build_check/ClickHouse/docker/packager && ./packager --output-dir=/home/ubuntu/actions-runner/_work/_temp/build_check/package_release --package-type=deb --compiler=clang-14 --cache=ccache --ccache_dir=/home/ubuntu/actions-runner/_work/_temp/../ccaches/package_release_ccache --additional-pkgs --docker-image-version=latest --version=22.6.1.1789 --official
2022-06-13T14:34:47.2814360Z 2022-06-13 14:34:47,280 Will build ClickHouse pkg with cmd: 'docker run --network=host --user=1000:1000 --rm --volume=/home/ubuntu/actions-runner/_work/_temp/build_check/package_release:/output --volume=/home/ubuntu/actions-runner/_work/_temp/build_check/ClickHouse/docker/packager/../..:/build --volume=/home/ubuntu/actions-runner/_work/_temp/../ccaches/package_release_ccache:/ccache  -e OUTPUT_DIR=/output -e DEB_ARCH=amd64 -e MAKE_DEB=true -e WITH_PERFORMANCE=1 -e BUILD_MUSL_KEEPER=1 -e CC=clang-14 -e CXX=clang++-14 -e BUILD_TYPE=None -e CCACHE_DIR=/ccache -e CCACHE_BASEDIR=/build -e CCACHE_NOHASHDIR=true -e CCACHE_COMPILERCHECK=content -e CCACHE_MAXSIZE=15G -e MAKE_APK=true -e MAKE_RPM=true -e MAKE_TGZ=true -e VERSION_STRING='22.6.1.1789' -e AUTHOR='clickhouse' -e CMAKE_FLAGS="$CMAKE_FLAGS -DENABLE_TESTS=0 -DENABLE_UTILS=0 -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_AUTOGEN_VERBOSE=ON -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DINSTALL_STRIPPED_BINARIES=ON -DCMAKE_C_COMPILER=clang-14 -DCMAKE_CXX_COMPILER=clang++-14 -DCLICKHOUSE_OFFICIAL_BUILD=1"  clickhouse/binary-builder:latest'
```

It's master. In the PR builds, `official` is not there indeed. And there's the reason why: https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/build_check.py#L253

The flag is there only for builds from master/release PRs

rschu1ze commented:
Thanks, Misha. I am still confused, sorry.
(1) This PR is a merge from branch "ClickHouse/stripping" into "ClickHouse/master"
(2) [Maksims PR](https://github.com/ClickHouse/ClickHouse/pull/38023) is a merge from "kitaisreal:examples-coroutines-build-fix" into "ClickHouse/master".

Compared to (1), (2) has no special labels, they are merging into the same repo/branch, and (2) does not provide a special "pr_info.number" (checked by build_check.py). So why does (2) produce an official build but (1) doesn't? And what can I do in this PR to create an official build?

rschu1ze committed:
Minor follow-up

- change ELF section name to ".clickhouse.hash" (lowercase seems
  standard)

- more expressive/concise integrity check messages at startup

Felixoid commented:
Sorry, but I can't find any build in Max's PR. They both failed because of the first CheckLabels.

In short, we only build official binaries from `master` OR release PRs, e.g. #37345 and #36501 

You can temporary tweak this part in the PR https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/build_check.py#L253 to force it

rschu1ze committed:
(Temporary / to be reverted) Force an official build.

rschu1ze commented:
Nice, thanks!

In the meantime I figured out that my attempt to rename ELF section `.note.ClickHouse.hash` to `.clickhouse.hash` so it would survive  `strip --remove-section=.comment --remove-section=.note` didn't work. `Strip` killed the hash regardless ... What does work is adding `--keep-section=.clickhouse.hash`...

rschu1ze committed:
Retain .clickhouse.hash when stripping

mergify[bot] committed:
Merge branch 'master' into stripping

rschu1ze commented:
With forced "official" builds (= forced hash embedding + forced stripping), I am now seeing a successful integrity check in all CI tests. I.e. the fix works as expected.

Will revert the "force official build" commit. Afterwards, we can submit. Would be happy if @Felixoid or @alexey-milovidov could approve. Tnx!

rschu1ze committed:
Revert "(Temporary / to be reverted) Force an official build."

This reverts commit 45d309108781842876baa8e4bd51827eb242bcaf.

Felixoid commented on src/Daemon/BaseDaemon.cpp:
```diff
@@ -355,23 +355,23 @@ class SignalListener : public Poco::Runnable
         {
-            LOG_FATAL(log, "Calculated checksum of the binary: {}."
-                " There is no information about the reference checksum.", calculated_binary_hash);
+            LOG_FATAL(log, "Integrity check of the executable skipped because the reference checksum could not be read."
```
Just a question if it makes sense to reuse this in programs/server/Server.cpp

rschu1ze replied:
Yeah, there is some copypasta. Reuse is probably not worth the trouble as its in two files but what I can do is to add a comment to keep the text at least in sync.

Felixoid reviewed (approved):
Besides a tiny note - good to merge :+1:

rschu1ze committed:
+ comments about keeping stuff in sync
